### PR TITLE
New profile: Quodlibet

### PR DESCRIPTION
**Pull request by Bundy01**

### CODE REVIEW

File: etc/profile-m-z/quodlibet.profile
Comment (rusty-snake) [inline etc/profile-m-z/quodlibet.profile:12]: Add `blacklist`s in disable-programs.inc

File: etc/profile-m-z/quodlibet.profile
Comment (rusty-snake) [inline etc/profile-m-z/quodlibet.profile:17]: - Add `mkdir`/`mkfile`s
- Add `include whitelist-common.inc`
- Move the whitelist down (between `include disable-*.inc` and `apparmor`)
- Can we add `whitelist-runuser-common.inc`, `whitelist-var-common.inc`, `whitelist-usr-share-common.inc`?

File: etc/profile-m-z/quodlibet.profile
Comment (rusty-snake) [inline etc/profile-m-z/quodlibet.profile:31]: That's a feh specific include, add `netfilter`, `protocol` and the `private-etc` file directly.

File: etc/profile-m-z/quodlibet.profile
Comment (rusty-snake) [inline etc/profile-m-z/quodlibet.profile:51]: Does it need D-Bus access or can be add `dbus-{user,system} none`?

File: etc/profile-m-z/quodlibet.profile
Comment (rusty-snake) [inline etc/profile-m-z/quodlibet.profile:35]: Breaks pulseaudio, remove.

File: etc/profile-m-z/quodlibet.profile
Comment (rusty-snake) [inline etc/profile-m-z/quodlibet.profile:43]: Can we add `seccomp.block-secondary`?

File: etc/profile-m-z/quodlibet.profile
Comment (rusty-snake) [inline etc/profile-m-z/quodlibet.profile:47]: Run sort.py

File: etc/profile-m-z/quodlibet.profile
Comment (rusty-snake) [inline etc/profile-m-z/quodlibet.profile:29]: Does it need write access or can we `include disable-write-mnt.inc` and `read-only ${HOME}`?

**Review by rusty-snake — COMMENTED** [review]
Looks good so far, some nitpicks and future hardening suggestion below.

It should also be added to firecfg.config so that firecfg creates symlinks for it.

**Comment by Bundy01** [comment]
add read-only ${HOME} ?
I don't know.

Breaks pulseaudio, remove.
What should I remove, because on my side it's functional.

Does it need D-Bus access or can be add dbus-user none?
Quodlibet needs to communicate with my DE (Cinnamon).

File: etc/profile-m-z/quodlibet.profile
Comment (rusty-snake) [reply etc/profile-m-z/quodlibet.profile:35]: > What should I remove, because on my side it's functional.

`machine-id`

File: etc/profile-m-z/quodlibet.profile
Comment (rusty-snake) [reply etc/profile-m-z/quodlibet.profile:51]: > Quodlibet needs to communicate with my DE (Cinnamon).

`dbus-system none` should work then, if it's just the DE.

File: etc/profile-m-z/quodlibet.profile
Comment (rusty-snake) [reply etc/profile-m-z/quodlibet.profile]: `seccomp.block-secondary` does not imply `seccomp`.

File: etc/profile-m-z/quodlibet.profile
Comment (rusty-snake) [inline etc/profile-m-z/quodlibet.profile:59]: add `crypto-policies`

File: etc/profile-m-z/quodlibet.profile
Comment (rusty-snake) [inline etc/profile-m-z/quodlibet.profile:25]: https://github.com/netblue30/firejail/blob/4b2914eb554635ce11e0b721c59cbb0ec428af44/etc/templates/profile.template#L28-L33

**Comment by Bundy01** [comment]
Sorry, I don't know how to use sort.py :(

**Comment by rusty-snake** [comment]
`python3 /path/to/sort.py /path/to/quodlibet.profile`

However this gave me the idea to print the fixed lines when running in the check in GitHub Actions, so everyone can just copy it from there if they can't use it for any reasons. PR will come.

**Comment by Bundy01** [comment]
> python3 /path/to/sort.py /path/to/quodlibet.profile

Thanks.

**Comment by glitsj16** [comment]
More a reminder note than anything else. We already have a profile for [exfalso](https://github.com/netblue30/firejail/blob/master/etc/profile-a-l/exfalso.profile), which is part of the quodlibet sources IIRC. In that context it might be worthwhile to consider refactoring this (or exfalso.profile) as a redirect. If it's too much trouble in this PR, we can always look at it later.

**Comment by Bundy01** [comment]
Hello,
sorry for the late response. I can try to rewrite Exfalso's profile which redirects to Quodlibet but it will take me a few days to test the changes.

**Comment by glitsj16** [comment]
@Bundy01 Great! Take your time. I've been using exfalso for ages and can always test from this side to assist.

**Comment by Bundy01** [comment]
I would like to add `private-lib` in quodlibet's profile but I'm breaking my teeth on it. Help would be welcome, otherwise I finished the profile.

**Comment by glitsj16** [comment]
> I would like to add `private-lib` in quodlibet's profile but I'm breaking my teeth on it. Help would be welcome, otherwise I finished the profile.

Can you show us what you've put together already? Sidenote: `private-lib` is notoriously difficult with media players. They can have a busload of extra plugins to support extra formats etcetera. IMO that's the main reason why none of our player apps like amarok, ffmpeg, mpv, mplayer, vlc, totem, ... has `private-lib`. So don't worry about it too much. But we can give it a try.

**Comment by rusty-snake** [comment]
Quodlibet is written in python. We can not add `private-lib` until #3236 is fixed, because it will not work for every system/distro.

**Comment by Bundy01** [comment]
Ok, so much for 'private-lib'.
Here are the profiles I modified, if it suits you I do the PR.

```
# Firejail profile for quodlibet
# Description: Music player and music library manager
# This file is overwritten after every install/update
# Persistent local customizations
include quodlibet.local
# Persistent global definitions
include globals.local

noblacklist ${MUSIC}

# Allow python (blacklisted by disable-interpreters.inc)
include allow-python2.inc
include allow-python3.inc

include disable-common.inc
include disable-devel.inc
include disable-exec.inc
include disable-interpreters.inc
include disable-passwdmgr.inc
include disable-programs.inc
include disable-write-mnt.inc
include disable-xdg.inc

mkdir ${HOME}/.cache/quodlibet
mkdir ${HOME}/.config/quodlibet
mkdir ${HOME}/.quodlibet

whitelist ${MUSIC}
whitelist ${HOME}/.cache/quodlibet
whitelist ${HOME}/.config/quodlibet
whitelist ${HOME}/.quodlibet

include whitelist-common.inc
include whitelist-runuser-common.inc
include whitelist-usr-share-common.inc
include whitelist-var-common.inc

apparmor
caps.drop all
ipc-namespace
netfilter
no3d
nodvd
nogroups
nonewprivs
noroot
notv
nou2f
novideo
protocol unix,inet,inet6
seccomp
seccomp.block-secondary
shell none
tracelog

private-bin exfalso,operon,python*,quodlibet,sh
private-cache
private-dev
private-etc alsa,alternatives,asound.conf,ca-certificates,crypto-policies,dconf,fonts,gtk-3.0,ld.so.cache,ld.so.conf,ld.so.conf.d,ld.so.preload,pki,pulse,resolv.conf,ssl
private-tmp

dbus-system none

#memory-deny-write-execute - breaks on Arch (see issue #1803)
```


```
# Firejail profile for exfalso
# Description: GTK audio tag editor
# This file is overwritten after every install/update
# Persistent local customizations
include exfalso.local
# Persistent global definitions
include globals.local

private-lib libatk-1.0.so.*,libgdk-3.so.*,libgdk_pixbuf-2.0.so.*,libgirepository-1.0.so.*,libgstreamer-1.0.so.*,libgtk-3.so.*,libgtksourceview-3.0.so.*,libpango-1.0.so.*,libpython*,libreadline.so.*,libsoup-2.4.so.*,libssl.so.1.*,python2*,python3*

dbus-user none

include quodlibet.profile

```

**Comment by rusty-snake** [comment]
- Remove `ipc-namespace`, it may cause issues under X11
- The mdwe note can be removed too.
- We could add `include allow-bin-sh.inc` with `include disable-shell.inc`  too.

File: etc/profile-m-z/quodlibet.profile
Comment (Bundy01) [reply etc/profile-m-z/quodlibet.profile:24]: read-only ${HOME} break extensions.

Comment (rusty-snake) [reply etc/profile-m-z/quodlibet.profile:24]: Does it work if you add `read-write ${HOME}/.quodlibet`?

Comment (Bundy01) [reply etc/profile-m-z/quodlibet.profile:24]: I didn't try, but extensions modify the files in the MUSIC directory.

Comment (rusty-snake) [reply etc/profile-m-z/quodlibet.profile:24]: If it need to modify files, remove the `read-only ${HOME}`. We have already `whitelist`ing.

File: etc/profile-a-l/exfalso.profile
Comment (rusty-snake) [inline etc/profile-a-l/exfalso.profile:13]:
```suggestion
# Redirect
include quodlibet.profile
```

File: etc/profile-a-l/exfalso.profile
Comment (rusty-snake) [inline etc/profile-a-l/exfalso.profile:7]:
```suggestion
# added by included profile
#include globals.local
```

Maybe we should `ignore include disable-write-mnt.inc`.

File: etc/profile-m-z/quodlibet.profile
Comment (rusty-snake) [inline etc/profile-m-z/quodlibet.profile:30]:
```suggestion
whitelist ${MUSIC}
whitelist ${DOWNLOADS}
```

Comment (Bundy01) [reply etc/profile-m-z/quodlibet.profile:30]: I liked the idea of allowing only the MUSIC folder, but I'll make the changes.

Comment (rusty-snake) [reply etc/profile-m-z/quodlibet.profile:30]: I remarked it because exfalso had it, and without it we break something that was working w/o significant security win. Alternative we could add/keep `whitelist ${DOWNLOADS}` only in exfalso.

**Review by glitsj16 — CHANGES_REQUESTED** [review]

File: etc/profile-a-l/exfalso.profile
Comment (Bundy01) [reply etc/profile-a-l/exfalso.profile:54]: Hi,
for private-etc, I quickly tested exfalso without any apparent problem, hence my modification. If I need to, I can add the problem line.
Is there anything else to expect because I would like it to be the last PR :)
Thanks for the answer.

**Comment by rusty-snake** [comment]
@Bundy01 @glitsj16 what's the state here?

**Comment by Bundy01** [comment]
@rusty-snake:
On my side, it's OK.

File: etc/profile-m-z/quodlibet.profile
Comment (glitsj16) [inline etc/profile-m-z/quodlibet.profile:8]: Add:
noblacklist ${HOME}/.cache/quodlibet
noblacklist ${HOME}/.config/quodlibet
noblacklist ${HOME}/.quodlibet

The first two need to be added as blacklists to disable-programs.inc (blacklist ${HOME}/.quodlibet is aleady in that file).

File: etc/profile-m-z/quodlibet.profile
Comment (glitsj16) [inline etc/profile-m-z/quodlibet.profile:30]: Put whitelist ${DOWNLOADS} above whitelist ${MUSIC} to keep things alphabetically ordered.

**Review by glitsj16 — CHANGES_REQUESTED** [review]
See my newly added comments.

**Comment by Bundy01** [comment]
Hi,
I was not able to use sort.py with the command: `python3 /path/to/sort.py /path/to/profile` :(

File: etc/profile-m-z/quodlibet.profile
Comment (rusty-snake) [inline etc/profile-m-z/quodlibet.profile:15]:
```suggestion
include allow-bin-sh.inc

# Allow python (blacklisted by disable-interpreters.inc)
```

File: etc/profile-m-z/quodlibet.profile
Comment (rusty-snake) [inline etc/profile-m-z/quodlibet.profile:32]:
```suggestion
```

File: etc/profile-m-z/quodlibet.profile
Comment (rusty-snake) [inline etc/profile-m-z/quodlibet.profile:38]:
```suggestion
```

**Review by rusty-snake — COMMENTED** [review]
Just some blank lines.

> I was not able to use sort.py with the command: `python3 /path/to/sort.py /path/to/profile` :(

What was the error? What shows `python3 --version`? However the sort.py CI is passing for your latest commit.

File: etc/profile-m-z/quodlibet.profile
Comment (Bundy01) [reply etc/profile-m-z/quodlibet.profile:38]: Do you really want me to delete line 38.There would no longer be a space between the `whitelist` and `include` values?

Comment (rusty-snake) [reply etc/profile-m-z/quodlibet.profile:38]: There is no blank line in any other profile (except for them where this is wrong too).

**Review by glitsj16 — APPROVED** [review]
Unblocking this so it can be merged when @rusty-snake is happy with it. We can always work out any kinks later.

File: etc/profile-m-z/quodlibet.profile
Comment (Bundy01) [reply etc/profile-m-z/quodlibet.profile]: @rusty-snake: Do you suggest I remove the space between `mkdir` and `whitelist`?

Comment (rusty-snake) [reply etc/profile-m-z/quodlibet.profile]: yes.

Do it now or I merge it anyway.

**Comment by rusty-snake** [comment]
@glitsj16 FYI you can dismiss your review if you don't want to block a PR but also don't want to approve it.
IDK why this is hidden behind 3 clicks.

Screenshot_2021-04-14_16-27-14-fs8

**Comment by rusty-snake** [comment]
merged, thanks.

**Comment by glitsj16** [comment]
> FYI you can dismiss your review if you don't want to block a PR but also don't want to approve it. IDK why this is hidden behind 3 clicks.

@rusty-snake Aha, I didn't see that option. Good to know, thanks for the info! Glad to see this merged, now I can start to double-check all my local overrides for exfalso. Thanks @Bundy01 for pursuing this!

**Pull request merged**
